src/Tuple/index.js

```diff
@@ -0,0 +1,152 @@
+/** @license ISC License (c) copyright 2016 original and current authors */
+/** @author Karthik Iyengar (karthikiyengar) */
+
+const VERSION = 1
+
+const _implements = require('../core/implements')
+const _equals = require('../core/equals')
+const _inspect = require('../core/inspect')
+const _type = require('../core/types').type('Tuple')
+const typeString = require('../core/types').typeFn(_type(), VERSION)
+const fl = require('../core/flNames')
+
+const isFunction = require('../core/isFunction')
+const isInteger = require('../core/isInteger')
+const isSameType = require('../core/isSameType')
+const isSemigroup = require('../core/isSemigroup')
+
+const constant = x => () => x
+
+function _Tuple(n) {
+  if (!isInteger(n) || n < 1 || n > 10) {
+    throw new TypeError(
+      'Tuple: Tuple size should be a number between 1 and 10'
+    )
+  }
+
+  /* eslint-disable no-unused-vars */
+  switch (n) {
+  case 1: return function(a) { return Tuple(n, arguments) }
+  case 2: return function(a, b) { return Tuple(n, arguments) }
+  case 3: return function(a, b, c) { return Tuple(n, arguments) }
+  case 4: return function(a, b, c, d) { return Tuple(n, arguments) }
+  case 5: return function(a, b, c, d, e) { return Tuple(n, arguments) }
+  case 6: return function(a, b, c, d, e, f) { return Tuple(n, arguments) }
+  case 7: return function(a, b, c, d, e, f, g) { return Tuple(n, arguments) }
+  case 8: return function(a, b, c, d, e, f, g, h) { return Tuple(n, arguments) }
+  case 9: return function(a, b, c, d, e, f, g, h, i) { return Tuple(n, arguments) }
+  case 10: return function(a, b, c, d, e, f, g, h, i, j) { return Tuple(n, arguments) }
+  }
+  /* eslint-enable no-unused-vars */
+
+  function Tuple(n, args) {
+    const parts = [].slice.call(args)
+    if (n !== parts.length) {
+      throw new TypeError(
+        `${n}-Tuple: Expected ${n} values, but got ${parts.length}`
+      )
+    }
+
+    const inspect = () => `Tuple(${parts.map(_inspect).join(',')} )`
+
+    const type = constant(`${n}-${_type()}`)
+
+    function map(method) {
+      return function(fn) {
+        if (!isFunction(fn)) {
+          throw new TypeError(`${n}-Tuple.${method}: Function required`)
+        }
+        return _Tuple(n)(
+          ...parts.slice(0, parts.length - 1).concat(fn(parts[parts.length - 1]))
+        )
+      }
+    }
+
+    function equals(m) {
+      return isSameType({ type }, m)
+      && _equals(parts, m.toArray())
+    }
+
+    function concat(method) {
+      return function(t) {
+        if (!isSameType({ type }, t)) {
+          throw new TypeError(`${n}-Tuple.${method}: Tuple of the same length required`)
+        }
+        const a = t.toArray()
+        return _Tuple(n)(
```

Comment:
Do we need to go through `_Tuple` for 🌽struction here? All other functions seem to use `Tuple`